edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/CertificateRenewal.cs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+
+namespace Microsoft.Azure.Devices.Edge.Hub.Service
+{
+    using System;
+    using System.Threading;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class CertificateRenewal : IDisposable
+    {
+        readonly CancellationTokenSource cts;
+
+        public CancellationToken Token => this.cts.Token;
+
+        public CertificateRenewal(EdgeHubCertificates certificates, ILogger logger)
+        {
+            Preconditions.CheckNotNull(certificates, nameof(certificates));
+            Preconditions.CheckNotNull(logger, nameof(logger));
+
+            TimeSpan timeToExpire = certificates.ServerCertificate.NotAfter - DateTime.UtcNow;
+            if (timeToExpire > TimeSpan.Zero)
+            {
+                logger.LogInformation("Scheduling server certificate renewal for {0}.", DateTime.UtcNow.Add(timeToExpire).ToString("o"));
+                this.cts = new CancellationTokenSource(timeToExpire);
+                this.cts.Token.Register(l => ((ILogger)l).LogInformation("Performing server certificate renewal."), logger);
```

Comment:
Nit: We should probably also add "restarting edgeHub.." to the log..

Reply:
+1. Perhaps "Restarting Edge for certificate renewal..

Reply:
I think I can add a log in main to indicate we are shutting down. However, most of the logs are in ShutdownHandler and guarded by a semaphore. Do you know the history here?

Reply:
I added a couple logs to main?

Reply:
Cool, that works...